### PR TITLE
Flop Counter Improvements

### PR DESCRIPTION
**Pull request by sebwolf-de**
* Write Flop Counter per node to csv file for detailed analysis.
* Compute current performance (since last syncpoint) instead of average since the beginning. This way you better see changes in performance.
* Better output to stdout.

### CODE REVIEW

**Comment by krenzland** [comment]
> Compute current performance (since last syncpoint) instead of average since the beginning. This way you better see changes in performance.

Is this in addition to the old way of outputting or replacing it?

I agree that it is valuable information, however, I **need** the overall FLOPS as well. Otherwise reporting the performance of unfinished runs is impossible. This feature is really needed for full-machine runs for example.

File: src/Solver/Interoperability.cpp [hunk 1]
```diff
@@ -975,6 +975,8 @@ seissol::Interoperability::initializeIO(int numSides, int numBndGP, int refineme
                     freeSurfaceFilename,
                     energySyncInterval);
 
+  seissol::SeisSol::main.flopCounter().init(freeSurfaceFilename);
```
Comment (Thomas-Ulrich) [inline src/Solver/Interoperability.cpp:978]: Should  it be named outputFileNamePrefix?

**Comment by sebwolf-de** [comment]
It's a replacement, but we can also output both.

File: src/Monitoring/FlopCounter.hpp [hunk 2]
```diff
@@ -41,24 +41,37 @@
...
+  void init(std::string outputFileNamePrefix);
+  void printPerformanceUpdate(double wallTime);
+  void printPerformanceSummary(double wallTime);
 
-void printPerformance(double wallTime);
-void printFlops();
+  private:
+  std::ofstream out;
+  long long recentTotalFlops = 0;
+  double recentWallTime = 0;
```
Comment (Thomas-Ulrich) [inline src/Monitoring/FlopCounter.hpp:73]: what about previousCallWallTime and previousCallTotalFlops?
or previousWallTime?

**Review by Thomas-Ulrich — COMMENTED** [review]
Looks good. I also tested and it works.

**Comment by krenzland** [comment]
Yes, please do both.

**Review by krenzland — CHANGES_REQUESTED** [review]
LGTM
However:
GFLOP/s -> GFLOPS (standard nomenclature) 
Please output both recent and overall GFLOPS

File: src/Monitoring/FlopCounter.hpp [hunk 3]
```diff
@@ -41,24 +41,37 @@
...
-extern long long g_SeisSolNonZeroFlopsNeighbor;
-extern long long g_SeisSolHardwareFlopsNeighbor;
-extern long long g_SeisSolNonZeroFlopsOther;
-extern long long g_SeisSolHardwareFlopsOther;
-extern long long g_SeisSolNonZeroFlopsDynamicRupture;
-extern long long g_SeisSolHardwareFlopsDynamicRupture;
-extern long long g_SeisSolNonZeroFlopsPlasticity;
-extern long long g_SeisSolHardwareFlopsPlasticity;
+namespace seissol::monitoring {
+class FlopCounter {
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply src/Monitoring/FlopCounter.hpp:52]: Actually, I want `previousTotalFlops` and `previousWallTime` to be private.

[the remark this answers is not in the document]
Comment (krenzland) [reply src/Monitoring/FlopCounter.hpp]: You can have private members in structs

[the remark this answers is not in the document]
Comment (krenzland) [reply src/Monitoring/FlopCounter.hpp]: Actually, struct/class are identical, the only difference is default visibility

**Review by krenzland — APPROVED** [review]

**Pull request merged**